File: ai-interview/frontend/src/api/experiences.ts
import { apiClient } from './client';
import type {
  ExperienceDetail,
  ExperienceImportPayload,
  ExperienceImportWebPayload,
  ExperienceListItem,
  ExperienceSearchPayload,
  ExperienceSearchResponse,
  ExperienceUpdatePayload,
} from '../types/experience';

export interface ExperienceFilters {
  target_school?: string;
  target_major?: string;
  interview_type?: string;
  source_type?: string;
}

export async function listExperiences(filters: ExperienceFilters = {}) {
  const { data } = await apiClient.get<ExperienceListItem[]>('/experiences', { params: filters });
  return data;
}

export async function getExperience(experienceId: string) {
  const { data } = await apiClient.get<ExperienceDetail>(`/experiences/${experienceId}`);
  return data;
}

export async function importExperienceText(payload: ExperienceImportPayload) {
  const { data } = await apiClient.post<{ experience_id: string; extract_status: string }>('/experiences/import-text', payload);
  return data;
}

export async function searchExperienceWeb(payload: ExperienceSearchPayload) {
  const { data } = await apiClient.post<ExperienceSearchResponse>('/experiences/search-web', payload);
  return data;
}

export async function importExperienceWeb(payload: ExperienceImportWebPayload) {
  const { data } = await apiClient.post<{ experience_id: string; extract_status: string }>('/experiences/import-web', payload);
  return data;
}

export async function updateExperience(experienceId: string, payload: ExperienceUpdatePayload) {
  const { data } = await apiClient.patch<ExperienceDetail>(`/experiences/${experienceId}`, payload);
  return data;
}

export async function deleteExperience(experienceId: string) {
  await apiClient.delete(`/experiences/${experienceId}`);
}

export async function extractExperience(experienceId: string, llm_config_id?: string | null) {
  const { data } = await apiClient.post<ExperienceDetail>(`/experiences/${experienceId}/extract`, { llm_config_id: llm_config_id || null });
  return data;
}

File: ai-interview/backend/app/api/routes/experiences.py
from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.user import User
from app.schemas.experience import (
    ExperienceDetail,
    ExperienceExtractRequest,
    ExperienceImportResponse,
    ExperienceImportTextRequest,
    ExperienceImportWebRequest,
    ExperienceListItem,
    ExperienceSearchWebRequest,
    ExperienceSearchWebResponse,
    ExperienceUpdateRequest,
)
from app.services.ai.llm_service import LLMService
from app.services.experience_service import ExperienceService

router = APIRouter(prefix="/experiences", tags=["experiences"])


def get_llm_service() -> LLMService:
    return LLMService()


@router.get("", response_model=list[ExperienceListItem])
def list_experiences(
    target_school: str | None = Query(default=None),
    target_major: str | None = Query(default=None),
    interview_type: str | None = Query(default=None),
    source_type: str | None = Query(default=None),
    db: Session = Depends(get_db),
    llm: LLMService = Depends(get_llm_service),
    current_user: User = Depends(get_current_user),
):
    return ExperienceService(db, llm).list_for_user(
        current_user.id,
        target_school=target_school,
        target_major=target_major,
        interview_type=interview_type,
        source_type=source_type,
    )


@router.post("/import-text", response_model=ExperienceImportResponse)
async def import_experience_text(
    payload: ExperienceImportTextRequest,
    db: Session = Depends(get_db),
    llm: LLMService = Depends(get_llm_service),
    current_user: User = Depends(get_current_user),
):
    return await ExperienceService(db, llm).import_text(payload, current_user.id)


@router.post("/search-web", response_model=ExperienceSearchWebResponse)
async def search_web(
    payload: ExperienceSearchWebRequest,
    db: Session = Depends(get_db),
    llm: LLMService = Depends(get_llm_service),
    current_user: User = Depends(get_current_user),
):
    return await ExperienceService(db, llm).search_web(payload)


@router.post("/import-web", response_model=ExperienceImportResponse)
async def import_experience_web(
    payload: ExperienceImportWebRequest,
    db: Session = Depends(get_db),
    llm: LLMService = Depends(get_llm_service),
    current_user: User = Depends(get_current_user),
):
    return await ExperienceService(db, llm).import_web(payload, current_user.id)


@router.get("/{experience_id}", response_model=ExperienceDetail)
def get_experience(
    experience_id: str,
    db: Session = Depends(get_db),
    llm: LLMService = Depends(get_llm_service),
    current_user: User = Depends(get_current_user),
):
    return ExperienceService(db, llm).get_detail(experience_id, current_user.id)


@router.patch("/{experience_id}", response_model=ExperienceDetail)
def update_experience(
    experience_id: str,
    payload: ExperienceUpdateRequest,
    db: Session = Depends(get_db),
    llm: LLMService = Depends(get_llm_service),
    current_user: User = Depends(get_current_user),
):
    return ExperienceService(db, llm).update(experience_id, current_user.id, payload)


@router.delete("/{experience_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_experience(
    experience_id: str,
    db: Session = Depends(get_db),
    llm: LLMService = Depends(get_llm_service),
    current_user: User = Depends(get_current_user),
):
    ExperienceService(db, llm).delete(experience_id, current_user.id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{experience_id}/extract", response_model=ExperienceDetail)
async def extract_experience(
    experience_id: str,
    payload: ExperienceExtractRequest,
    db: Session = Depends(get_db),
    llm: LLMService = Depends(get_llm_service),
    current_user: User = Depends(get_current_user),
):
    return await ExperienceService(db, llm).extract(experience_id, current_user.id, payload)
